CA2241: AutoDetectStringFormattingMethods

manfred-brands opened the pull request:
In the end this was very simple as the method to detect a Format method already existed.

We could completely delete the existing configuration and simply check if the called method has a format parameter followed by a params parameter. This can not be much more expensive than checking the configuration.

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/roslyn-analyzers?pullRequest=3653) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/roslyn-analyzers?pullRequest=3653) before we can accept your contribution.<br/><br/>:x: manfred-brands [sign now](https://cla.dotnetfoundation.org/dotnet/roslyn-analyzers?pullRequest=3653)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/roslyn-analyzers?pullRequest=3653) it.</sub>

mavasani commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ProvideCorrectArgumentsToFormattingMethods.cs:
```diff
@@ -355,6 +355,13 @@ public StringFormatInfo(Compilation compilation)
 
+                // Check if method could be a string formatting method.
+                // This is defined as a method with a 'format' argument followed by an 'params object[]'.
+                if (TryGetFormatInfo(method, out info))
```
As far as I recall, this behavior was considered in the original implementation, but was not chosen due to following reasons:
1. Potential of being a breaking change for the rule.
2. Makes this a heuristic based rule. Current implementation of the rule is conservative to the known or user configured set of formatting methods. This makes it aggressive to fire on all methods which have similar signature.

Given that this rule is marked as `RuleLevel.BuildWarningCandidate`, I am reluctant to change the default behavior to make it so aggressive. At the minimum, we need an option to guard this new behavior so users can select the preferred level of aggressiveness.

@dotnet/roslyn-analysis for opinions.

manfred-brands commented:
I added a commit which makes this behaviour configurable.
In addition the format parameter is now verified to be a string.

mavasani commented:
Change looks good to me overall with the default behavior being unchanged. Please add relevant documentation. Also we probably want a better option name, but I cannot think of a good one. Any suggestions?

manfred-brands commented:
Renamed option to: try_determine_additional_string_formatting_methods_automatically. Should be clear.
Added documentation and addressed your other review comments.

mavasani commented on docs/Analyzer Configuration.md:
```diff
@@ -239,6 +239,16 @@ Examples:
 |`dotnet_code_quality.additional_string_formatting_methods = NS.MyType.MyFormat(ParamType)` | Matches specific method 'MyFormat' with given fully qualified signature
 |`dotnet_code_quality.additional_string_formatting_methods = NS1.MyType1.MyFormat1(ParamType)\|NS2.MyType2.MyFormat2(ParamType)` | Matches specific methods 'MyFormat1' and 'MyFormat2' with respective fully qualified signature
 
+Option Name: `try_determine_additional_string_formatting_methods_automatically`
```
Can you please add a more detailed description here on what this option does?

manfred-brands replied:
Done.

mavasani commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ProvideCorrectArgumentsToFormattingMethods.cs:
```diff
@@ -355,6 +355,17 @@ public StringFormatInfo(Compilation compilation)
 
+                // Check if the user configured automatic determination of formatting methods.
+                // If so, check if the method called has a 'string format' parameter followed by an params array.
+                bool defaultTryDetermineAdditionalStringFormattingMethodsAutomatically = false;
+                var canTryDetermineAdditionalStringFormattingMethodsAutomatically = context.Options.GetBoolOptionValue(EditorConfigOptionNames.TryDetermineAdditionalStringFormattingMethodsAutomatically,
+                        Rule, context.Operation.Syntax.SyntaxTree, context.Compilation, defaultTryDetermineAdditionalStringFormattingMethodsAutomatically, context.CancellationToken);
```
```suggestion
                var determineAdditionalStringFormattingMethodsAutomatically = context.Options.GetBoolOptionValue(EditorConfigOptionNames.TryDetermineAdditionalStringFormattingMethodsAutomatically,
                        Rule, context.Operation.Syntax.SyntaxTree, context.Compilation, defaultValue: false, context.CancellationToken);
```

manfred-brands replied:
Done.

mavasani commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ProvideCorrectArgumentsToFormattingMethods.cs:
```diff
@@ -355,6 +355,17 @@ public StringFormatInfo(Compilation compilation)
+                bool defaultTryDetermineAdditionalStringFormattingMethodsAutomatically = false;
+                var canTryDetermineAdditionalStringFormattingMethodsAutomatically = context.Options.GetBoolOptionValue(EditorConfigOptionNames.TryDetermineAdditionalStringFormattingMethodsAutomatically,
+                        Rule, context.Operation.Syntax.SyntaxTree, context.Compilation, defaultTryDetermineAdditionalStringFormattingMethodsAutomatically, context.CancellationToken);
+                if (canTryDetermineAdditionalStringFormattingMethodsAutomatically &&
+                    TryGetFormatInfo(method, out info) && info.ExpectedStringFormatArgumentCount == -1)
```
```suggestion
                if (determineAdditionalStringFormattingMethodsAutomatically  &&
                    TryGetFormatInfo(method, out info) &&
                    info.ExpectedStringFormatArgumentCount == -1)
```

manfred-brands replied:
Done.

mavasani commented on src/Utilities/Compiler/Options/EditorConfigOptionNames.cs:
```diff
@@ -67,6 +67,13 @@ internal static partial class EditorConfigOptionNames
         public const string AdditionalStringFormattingMethods = "additional_string_formatting_methods";
 
+        /// <summary>
+        /// String option to enable heuristically detecting of additional string formatting methods for CA2241 (https://docs.microsoft.com/visualstudio/code-quality/ca2241).
```
This exact comment can be added to the `Analyzer Configuration.md` doc section.

manfred-brands replied:
I noticed that the comment is wrong, it should read 'Boolean option'.

manfred-brands commented:
Changes made and pushed. Would you like for me to collapse these commits into one or will that confuse the automatic system?

manfred-brands commented:
Thanks. Raised pull request for documentation: https://github.com/MicrosoftDocs/visualstudio-docs/pull/5391
